parser/Builder.cc

```diff
@@ -176,10 +176,15 @@ class Builder::Impl {
         return cond;
     }
 
-    void error(ruby_parser::dclass err, core::LocOffsets loc, std::string data = "") {
+    void error_without_recovery(ruby_parser::dclass err, core::LocOffsets loc, std::string data = "") {
         driver_->external_diagnostic(ruby_parser::dlevel::ERROR, err, loc.beginPos(), loc.endPos(), data);
     }
 
+    void error(ruby_parser::dclass err, core::LocOffsets loc, std::string data = "") {
+        auto range = ruby_parser::diagnostic::range(loc.beginPos(), loc.endPos());
+        driver_->diagnostics.emplace_back(ruby_parser::dlevel::ERROR, err, std::move(range), data);
+    }
+
```

Comment:
I don't have better names for these, but I am not sure that `error_without_recovery` is the best name here.  It makes me think that we're raising exceptions or similar.  (Unless we think that `error_with_recovery` is temporary and we're planning on driving it down, in which case, w/e)

Also, the magic of `external_diagnostic` is just gross -- not your fault, just commenting in passing.

Reply:
Yeah, my plan is to eliminate it. I have some work locally, and plan to keep iterating soon.